<a href="https://colab.research.google.com/github/Master-45-vic/CEG-ML-Competition/blob/prasanth_code_files/Smart_Grid_Power_Load_in_India.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Smart Grid / Power Load in India

## 1. Overview

India's power grid is one of the largest and most complex in the world. While thermal (coal) power provides the baseload, renewable sources like Hydroelectric and Nuclear power are critical for maintaining grid stability and reducing carbon emissions.

However, these sources have unique challenges:

Hydro Power is highly seasonal, depending heavily

*   Hydro Power is highly seasonal, depending heavily on monsoon rains and reservoir levels.
*   Nuclear Power is stable but requires precise scheduling for maintenance and fueling.



If grid operators can accurately forecast the generation capacity of these sources, they can better optimize the usage of coal and prevent power deficits (blackouts).

## 2. Goal

Goal is to build a Machine Learning **model that forecasts the Daily Power Generation** (in Million Units - MU) for Hydro and Nuclear power plants across India's five regional grids (Northern, Western, Southern, Eastern, North-Eastern).

## 3. Data Understanding

### 3.1 Load Data

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
df=pd.read_csv("/content/drive/MyDrive/CEG ML Competitiom/CEG ML Competition/Smart Grid  Power Load in India/train.csv", index_col='id')
display(df.head())

,date,source,Region,Generation
id,,,,
1,2013-03-31,Hydro,ER,30.0
2,2013-03-31,Hydro,NER,7.0
3,2013-03-31,Hydro,NR,139.0
4,2013-03-31,Hydro,SR,72.0
5,2013-03-31,Hydro,WR,43.0


### 3.2 Shape of Data

In [7]:
df.shape

(22995, 4)



*   No of Rows = 22995
*   No of Columns = 4




### Columns Description



*   id - Unique row identifier.
*   date - The date of the record.
*  source - The type of power plant:
  
         *  Hydro: Dependent on water levels/monsoon.
         *  Nuclear: Generally stable but subject to maintenance cycles.

*   Region - The power grid region:

         *   NR: Northern Region
         *   WR: Western Region
         *   SR: Southern Region
         *   ER: Eastern Region
         *   NER: North Eastern Region

*   Generation - (TARGET) The actual power generated in Million Units (MU).




### 3.3 Data Types of Each Columns

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22995 entries, 1 to 22995
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        22995 non-null  object 
 1   source      22995 non-null  object 
 2   Region      22995 non-null  object 
 3   Generation  22995 non-null  float64
dtypes: float64(1), object(3)
memory usage: 898.2+ KB


*   data,source and region are **object** data type
*   generation is **float** data type



### 3.4 Number of Unique Value in Each Columns

In [12]:
def unique_count(df):
  for i in df.columns:
    count = df[i].nunique()
    print(i, ": ", count)
unique_count(df)

date :  3044
source :  2
Region :  5
Generation :  387


### 3.5 Number of NaN Values

In [13]:
def NaN_count(df):
  for i in df.columns:
    count = df[i].isna().sum()
    print(i, ": ", count)
NaN_count(df)

date :  0
source :  0
Region :  0
Generation :  0


### 3.6 Duplicate Records

In [15]:
print("Number of Duplicate Records : ",df.duplicated().sum())

Number of Duplicate Records :  90
